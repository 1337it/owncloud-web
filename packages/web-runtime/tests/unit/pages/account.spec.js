import account from '../../../src/pages/account.vue'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import { createStore } from 'vuex-extensions'
import GetTextPlugin from 'vue-gettext'

const localVue = createLocalVue()
localVue.use(Vuex)

localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

localVue.directive('translate', {
  inserted: (el) => {}
})

const $route = {
  meta: {
    title: 'Some Title'
  }
}

const selectors = {
  pageTitle: '.oc-page-title',
  loaderStub: 'oc-loader-stub',
  accountPageTitle: '#account-page-title',
  editUrlButton: '[data-testid="account-page-edit-url-btn"]',
  editRouteButton: '[data-testid="account-page-edit-route-btn"]',
  accountPageInfo: '.account-page-info',
  groupInformation: '.account-page-info-groups'
}

describe('account', () => {
  describe('when the wrapper is still loading', () => {
    let wrapper
    beforeEach(() => {
      wrapper = getWrapper()
      wrapper.setData({ loading: true })
    })
    it('should render the loading state', () => {
      const pageTitle = wrapper.find(selectors.pageTitle)
      const ocLoader = wrapper.find(selectors.loaderStub)
      expect(pageTitle.text()).toBe($route.meta.title)
      expect(ocLoader.exists()).toBeTruthy()
    })
    it('should not render the account page content', () => {
      const accountPageTitle = wrapper.find(selectors.accountPageTitle)
      const accountPageInfo = wrapper.find('.account-page-info')
      expect(accountPageTitle.exists()).toBeFalsy()
      expect(accountPageInfo.exists()).toBeFalsy()
    })
  })
  describe('when the wrapper is not loading anymore', () => {
    it('should not render the loading state', async () => {
      const store = getStore()
      const wrapper = getWrapper(store)
      await wrapper.setData({ loading: false })
      const pageTitle = wrapper.find(selectors.pageTitle)
      const ocLoader = wrapper.find(selectors.loaderStub)
      expect(pageTitle.text()).toBe('Account')
      expect(ocLoader.exists()).toBeFalsy()
    })

    describe('edit buttons', () => {
      describe('edit url button', () => {
        it('should be displayed if not running with ocis', async () => {
          const store = getStore({
            server: 'http://server/address/',
            isOcis: false
          })
          const wrapper = getWrapper(store)
          await wrapper.setData({ loading: false })
          const editUrlButton = wrapper.find(selectors.editUrlButton)
          const editRouteButton = wrapper.find(selectors.editRouteButton)
          expect(editUrlButton).toMatchSnapshot()
          expect(editUrlButton.attributes()).toMatchSnapshot()
          expect(editRouteButton.exists()).toBeFalsy()
        })
        it('should not be displayed if running with ocis', async () => {
          const store = getStore({
            server: 'http://server/address/',
            isOcis: true
          })
          const wrapper = getWrapper(store)
          await wrapper.setData({ loading: false })
          const editButton = wrapper.find(selectors.editUrlButton)
          expect(editButton.exists()).toBeFalsy()
        })
      })
      describe('edit route button', () => {
        it('should be displayed if running with ocis and has navItems', async () => {
          const store = getStore({
            isOcis: true,
            getNavItemsByExtension: jest.fn(() => [{ route: 'some-route' }])
          })
          const wrapper = getWrapper(store)
          await wrapper.setData({ loading: false })
          const editRouteButton = wrapper.find(selectors.editRouteButton)
          expect(editRouteButton).toMatchSnapshot()
          expect(editRouteButton.attributes()).toMatchSnapshot()
        })
      })
    })

    describe('account information', () => {
      it('without group information', async () => {
        const store = getStore({
          user: {
            username: 'some-username',
            displayname: 'some-displayname',
            email: 'some-email'
          }
        })
        const wrapper = getWrapper(store)
        await wrapper.setData({ loading: false })

        const accountPageInfo = wrapper.find(selectors.accountPageInfo)
        expect(accountPageInfo).toMatchSnapshot()
      })
      it('with group information', async () => {
        const store = getStore()
        const wrapper = getWrapper(store)
        await wrapper.setData({ loading: false })
        await wrapper.setData({ groups: ['one', 'two', 'three'] })

        const groupInformation = wrapper.find(selectors.groupInformation)
        expect(groupInformation).toMatchSnapshot()
      })
    })
  })
})

function getWrapper(store = null) {
  const component = {
    ...account,
    mounted: jest.fn()
  }
  const opts = {
    localVue,
    mocks: {
      $route
    },
    stubs: {
      'oc-loader': true,
      'oc-button': true,
      'oc-icon': true
    }
  }
  if (store) opts.store = store
  return shallowMount(component, opts)
}

function getStore({
  user = {},
  server = '',
  getNavItemsByExtension = jest.fn(() => []),
  isOcis = false
} = {}) {
  return createStore(Vuex.Store, {
    getters: {
      user: () => user,
      configuration: () => ({
        server: server
      }),
      getNavItemsByExtension: () => getNavItemsByExtension,
      isOcis: () => isOcis
    }
  })
}
