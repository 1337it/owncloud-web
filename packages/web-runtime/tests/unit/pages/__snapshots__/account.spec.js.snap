// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account when the wrapper is not loading anymore account information with group information 1`] = `
<div class="account-page-info-groups oc-mb oc-width-1-2@s">
  <dt class="oc-text-normal oc-text-muted">
    Group memberships
  </dt>
  <dd><span>one, two, three</span></dd>
</div>
`;

exports[`account when the wrapper is not loading anymore account information without group information 1`] = `
<dl class="account-page-info oc-flex oc-flex-wrap">
  <div class="account-page-info-username oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Username</dt>
    <dd>
      some-username
    </dd>
  </div>
  <!---->
  <div class="account-page-info-displayname oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Display name</dt>
    <dd>
      some-displayname
    </dd>
  </div>
  <div class="account-page-info-email oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">Email</dt>
    <dd>some-email</dd>
  </div>
  <div class="account-page-info-groups oc-mb oc-width-1-2@s">
    <dt class="oc-text-normal oc-text-muted">
      Group memberships
    </dt>
    <dd><span>You are not part of any group</span></dd>
  </div>
</dl>
`;

exports[`account when the wrapper is not loading anymore edit buttons edit route button should be displayed if running with ocis and has navItems 1`] = `
<oc-button-stub variation="primary" type="router-link" to="some-route" data-testid="account-page-edit-route-btn">
  <oc-icon-stub name="edit"></oc-icon-stub>
  <translate-stub tag="span">Edit</translate-stub>
</oc-button-stub>
`;

exports[`account when the wrapper is not loading anymore edit buttons edit route button should be displayed if running with ocis and has navItems 2`] = `
Object {
  "data-testid": "account-page-edit-route-btn",
  "to": "some-route",
  "type": "router-link",
  "variation": "primary",
}
`;

exports[`account when the wrapper is not loading anymore edit buttons edit url button should be displayed if not running with ocis 1`] = `
<oc-button-stub variation="primary" type="a" href="http://server/address/index.php/settings/personal" data-testid="account-page-edit-url-btn">
  <oc-icon-stub name="edit"></oc-icon-stub>
  <translate-stub tag="span">Edit</translate-stub>
</oc-button-stub>
`;

exports[`account when the wrapper is not loading anymore edit buttons edit url button should be displayed if not running with ocis 2`] = `
Object {
  "data-testid": "account-page-edit-url-btn",
  "href": "http://server/address/index.php/settings/personal",
  "type": "a",
  "variation": "primary",
}
`;
